python/lsst/pipe/tasks/processCcdWithFakes.py

```diff
@@ -309,6 +326,111 @@ def run(self, fakeCat, exposure, wcs=None, photoCalib=None, exposureIdInfo=None)
+
+        Parameters
+        ----------
+        icSourceCat : `lsst.afw.table.SourceCatalog`
+            Catalog from which to copy fields.
+        sourceCat : `lsst.afw.table.SourceCatalog`
+            Catalog to which to copy fields.
+
+        Returns
+        -------
+        newCat : `lsst.afw.table.SourceCatalog`
+            Catalog which includes the copied fields.
+
+        The fields copied are those specified by `config.icSourceFieldsToCopy` that actually exist
+        in the schema of `icSourceCat`.
+
+        This version was based on and adapted from the one in calibrateTask.
+        """
+
+        # Make a new SourceCatalog with the data from sourceCat so that we can add the new columns to it
+        sourceSchemaMapper = afwTable.SchemaMapper(sourceCat.schema)
+        sourceSchemaMapper.addMinimalSchema(sourceCat.schema, True)
+
+        icSourceSchemaMapper = afwTable.SchemaMapper(icSourceCat.schema, sourceCat.schema)
+
+        # Add the desired columns from the config option icSourceFieldsToCopy
+        missingFieldNames = []
+        for fieldName in self.config.icSourceFieldsToCopy:
+            try:
+                schemaItem = icSourceCat.schema.find(fieldName)
+            except KeyError:
+                missingFieldNames.append(fieldName)
+            else:
+                # field found; if addMapping fails then raise an exception
+                icSourceSchemaMapper.editOutputSchema().addField(schemaItem.getField())
+                schema = icSourceSchemaMapper.editOutputSchema()
+                icSourceSchemaMapper.addMapping(schemaItem.getKey(), schema.find(fieldName).getField())
+
+        if missingFieldNames:
+            raise RuntimeError("isSourceCat is missing fields {} specified in "
+                               "icSourceFieldsToCopy".format(missingFieldNames))
+
+        self.calibSourceKey = icSourceSchemaMapper.addOutputField(afwTable.Field["Flag"]("calib_detected",
+                                                                  "Source was detected as an icSource"))
+
+        schema = icSourceSchemaMapper.getOutputSchema()
+        newCat = afwTable.SourceCatalog(schema)
+        newCat.reserve(len(sourceCat))
+        newCat.extend(sourceCat, sourceSchemaMapper)
+
+        # Set the aliases so it doesn't complain.
+        for k, v in sourceCat.schema.getAliasMap().items():
+            newCat.schema.getAliasMap().set(k, v)
+
+        mc = afwTable.MatchControl()
+        mc.findOnlyClosest = False  # return all matched objects
+        matches = afwTable.matchXy(newCat, icSourceCat, self.config.matchRadiusPix, mc)
+        deblendKey = newCat.schema["deblend_nChild"].asKey()
+        # if deblended, keep children
+        matches = [m for m in matches if m[1].get(deblendKey) == 0]
+
+        # Because we had to allow multiple matches to handle parents, we now
+        # need to prune to the best matches
```

Comment:
Why not prune the catalog before matching?

    select = newCat["deblend_nChild"] == 0
    matches = afwTable.matchXy(newCat[select], icSourceCat, self.config.matchRadiusPix)

That's a more efficient operation (fewer sources in the catalog) and less code.

Reply:
Because I'm silly and didn't really think about what I had copied. Thank you.